lib/db.c

```diff
@@ -11164,9 +11165,16 @@ grn_ctx_at(grn_ctx *ctx, grn_id id)
                 id,
                 name_size, name);
           }
-          grn_db_value_unlock(ctx, id, vp);
-          if (!success) {
-            goto exit;
+          if (grn_enable_reference_count) {
+            if (!success) {
+              grn_db_value_unlock(ctx, id, vp);
+              goto exit;
+            }
+          } else {
+            grn_db_value_unlock(ctx, id, vp);
+            if (!success) {
+              goto exit;
+            }
```

Comment:
This logic is difficult to understand. How about the following flow?

```c
if (!grn_db_value_wait(ctx, id, vp)) {
  /* set error */
  grn_db_value_unlock(ctx, id, vp); /* Always unlock on error */
  goto exit;
}
if (!grn_enable_reference_count) {
  grn_db_value_unlock(ctx, id, vp); /* Unlock on non reference count mode. */
}
```

Reply:
Thank you for your comment.
I've fixed at https://github.com/groonga/groonga/pull/1257/commits/37f884893e281ce777e5651de711f58c94be5959.